File: backend/app/services/device_service.py
from collections.abc import Sequence
from datetime import UTC, datetime, timedelta

from app.models.models import DeviceTrack
from app.repositories.device_repo import DeviceRepo
from app.services.base import BaseService


class DeviceService(BaseService):
    def __init__(self, repo: DeviceRepo):
        self.repo = repo

    async def get_owned_device(
        self, device_id: int, user_id: int
    ) -> DeviceTrack:
        """Return device if found and owned, otherwise raise."""
        device = await self.repo.get_device_track_by_id(device_id)
        return self.get_owned(device, user_id, "设备不存在", "无权访问此设备")

    async def get_device_by_id(self, device_id: int) -> DeviceTrack | None:
        return await self.repo.get_device_track_by_id(device_id)

    async def get_user_devices(self, user_id: int) -> Sequence[DeviceTrack]:
        return await self.repo.get_device_tracks_by_user_id(user_id)

    async def create_device(self, user_id: int, **data: str) -> DeviceTrack:
        return await self.repo.create_device_track(user_id, **data)

    async def update_device(
        self, device_id: int, **data: str
    ) -> DeviceTrack | None:
        return await self.repo.update_device_track(device_id, **data)

    async def delete_device(self, device_id: int) -> bool:
        return await self.repo.delete_device_track_by_id(track_id=device_id)

    async def update_device_status(
        self, device_id: int, status: str
    ) -> DeviceTrack | None:
        """更新设备状态"""
        return await self.repo.update_device_track(device_id, status=status)

    async def update_device_reminder_config(
        self, device_id: int, reminder_config: dict
    ) -> DeviceTrack | None:
        """更新设备提醒配置"""
        return await self.repo.update_device_track(
            device_id, reminder_config=reminder_config
        )

    async def get_upcoming_milestone_devices(
        self, user_id: int, days_ahead: int = 30
    ) -> Sequence[DeviceTrack]:
        """获取即将到达里程碑的设备（未来N天内）。

        里程碑计算逻辑属于业务规则，放在服务层而非仓库层。
        """
        devices = await self.repo.get_device_tracks_by_user_id(user_id)
        active_devices = [d for d in devices if d.status == "active"]

        today = datetime.now(UTC).date()
        future_date = today + timedelta(days=days_ahead)

        upcoming = []
        for device in active_devices:
            days_owned = (today - device.purchase_date.date()).days
            future_days = (future_date - device.purchase_date.date()).days

            reminder_config = device.reminder_config or {}
            milestones = reminder_config.get(
                "milestones", [100, 365, 730, 1000, 1825]
            )

            for milestone in milestones:
                if days_owned <= milestone <= future_days:
                    upcoming.append(device)
                    break

        return upcoming

File: frontend/src/services/deviceService.ts
import { deviceGateway } from "@/api/deviceGateway";
import type { Device, DeviceInput } from "@/api/deviceGateway";
import { useNotificationStore } from "@/stores/notification";

export type { Device, DeviceInput };

export interface DeviceService {
  getUserDevices: () => Promise<Device[]>;
  getDeviceById: (device_id: number) => Promise<Device>;
  createDevice: (data: DeviceInput) => Promise<Device>;
  updateDevice: (device_id: number, data: Omit<DeviceInput, "id">) => Promise<Device>;
  deleteDevice: (device_id: number) => Promise<void>;
  updateDeviceStatus: (device_id: number, status: "active" | "retired") => Promise<Device>;
  updateDeviceReminderConfig: (device_id: number, reminder_config: unknown) => Promise<Device>;
  testNotification: (device_id: number) => Promise<void>;
}

const gateway = deviceGateway;

export const deviceService: DeviceService = {
  async getUserDevices(): Promise<Device[]> {
    return gateway.getUserDevices();
  },

  async getDeviceById(device_id: number): Promise<Device> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    return gateway.getDeviceById(device_id);
  },

  async createDevice(data: DeviceInput): Promise<Device> {
    if (!data.name?.trim()) {
      return Promise.reject(new Error("设备名称不能为空"));
    }
    if (data.price < 0) {
      return Promise.reject(new Error("价格不能为负数"));
    }
    if (!["active", "retired"].includes(data.status)) {
      return Promise.reject(new Error("设备状态无效"));
    }
    return gateway.createDevice(data);
  },

  async updateDevice(device_id: number, data): Promise<Device> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    if (data.name !== undefined && !data.name.trim()) {
      return Promise.reject(new Error("设备名称不能为空"));
    }
    return gateway.updateDevice(device_id, data);
  },

  async deleteDevice(device_id: number): Promise<void> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    return gateway.deleteDevice(device_id);
  },

  async updateDeviceStatus(device_id: number, status): Promise<Device> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    if (!["active", "retired"].includes(status)) {
      return Promise.reject(new Error("设备状态无效"));
    }
    return gateway.updateDeviceStatus(device_id, status);
  },

  async updateDeviceReminderConfig(device_id: number, reminder_config: unknown): Promise<Device> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    return gateway.updateDeviceReminderConfig(device_id, reminder_config);
  },

  async testNotification(device_id: number): Promise<void> {
    if (!device_id || device_id <= 0) {
      return Promise.reject(new Error("无效的设备ID"));
    }
    return gateway.testNotification(device_id);
  },
};
